### PR TITLE
feat: added sqs message batching

### PR DESCRIPTION
**Pull request by krackerman**
#454

### CODE REVIEW

File: test/util/sqs.spec.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
+const path = require('path');
```
Comment (krackerman) [inline test/util/sqs.spec.js]: Some results in the expect statements can be a little large. Should we move some of the data to a mock folder to reduce file size?

File: src/util/sqs.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+const assert = require('assert');
```
Comment (krackerman) [inline src/util/sqs.js]: Exact copy from `stats-collector` only change is we are now passing through the `call` function

File: test/util/sqs.spec.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+          ]
+        );
+        nockDone();
+        done();
+      });
+    });
+
+    it('Testing empty messages', () => {
+      nockBack('send-message-batch-no-messages_recording.json', async (nockDone) => {
+        const result = await aws.sqs.sendMessageBatch(
```
Comment (krackerman) [inline test/util/sqs.spec.js]: Test may be redundant. Wanted to capture what happens when there are no messages

File: test/util/sqs.spec.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    aws = index();
+    nockBack.setMode('record');
+    nockBack.fixtures = path.join(__dirname, '..', '__cassettes', 'util', 'sqs');
+  });
+
+  after(() => helper.after());
+
+  describe('Testing sendMessageBatch', () => {
+    it('Testing send message success', (done) => {
+      nockBack('send-message-batch-success_recording.json', async (nockDone) => {
```
Comment (MacMcIrish) [inline test/util/sqs.spec.js]: Let's try to keep the recording naming consistent,
Rules are to keep the recording name and test name as close as possible and
`camelCaseName.json_recording.json`

File: test/util/sqs.spec.js (hunk 3 above)
Comment (MacMcIrish) [reply test/util/sqs.spec.js]: Not a bad test at all :) keep it in

File: src/util/sqs.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+module.exports = (call) => ({
+  sendMessageBatch: async (msgs, queueUrl, batchSize = 10) => {
+    const result = await Promise.all(chunk(msgs, batchSize)
+      .map((sqsBatch) => sendBatch(sqsBatch, queueUrl, call)));
+    assert(
+      msgs.length === result
+        .reduce((p, c) => p + c.reduce((prev, cur) => prev + cur.Successful.length, 0), 0),
+      `SendMessageBatch Error: ${abbrev(result)}\n`
+      + `Messages length: ${msgs.length}`
+    );
```
Comment (simlu) [inline src/util/sqs.js]: 1) no need for abbrev
2) Should throw explicitly (replace assert with throw)

File: test/util/sqs.spec.js (hunk 1 above)
Comment (MacMcIrish) [reply test/util/sqs.spec.js]: No, that shouldn't be necessary here. A better thing to do is to generify your data going out, as well as your expected results going back in. Think about this:

```js
const sentEvent = { type: 'discover', ... };

// success test
aws.sqs.sendMessageBatch([sentEvent]).to.deep.equal(...);

// fail test
aws.sqs.sendMessageBatch([sentEvent]).to.throw(...) // or w/e

File: test/mocha.opts [hunk 6]
```diff
@@ -1 +1 @@
---timeout 2000
+--timeout 10000
```
Comment (krackerman) [inline test/mocha.opts]: Will timeout when testing `SendMessageBatchError` as we retry `10` times with the exponential backoff function

File: src/util/sqs.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+      QueueUrl: queueUrl
+    });
+    response.push(result);
+    result.Successful.forEach((e) => delete pending[e.Id]);
+  }
+  return response;
+};
+
+module.exports = (call) => ({
+  sendMessageBatch: async (msgs, queueUrl, batchSize = 10) => {
```
Comment (simlu) [inline src/util/sqs.js]:
```suggestion
sendMessageBatch: async (msgs, queueUrl, {
  batchSize = 10,
  maxRetries = 10,
  backoffFunction = (count) => 30 * (count ** 2)
) => {
```

File: test/mocha.opts (hunk 6 above)
Comment (simlu) [reply test/mocha.opts]: revert and fix tests to run fast

File: test/resources/errors.spec.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+const expect = require('chai').expect;
```
Comment (simlu) [inline test/resources/errors.spec.js]: use node-tdd

File: test/util/sqs.spec.js [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+  });
+
+  desc('Testing sendMessageBatch', ({ it }) => {
+    it('Testing send message success', async () => {
+      const result = await aws.sqs.sendMessageBatch(
+        [
+          {
+            type: 'discover',
+            data:
+              {
```
Comment (simlu) [inline test/util/sqs.spec.js]: mv up

File: test/util/sqs.spec.js [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+  let aws;
+  before(() => {
+    aws = index();
+  });
+
+  desc('Testing sendMessageBatch', ({ it }) => {
+    it('Testing send message success', async () => {
+      const result = await aws.sqs.sendMessageBatch(
+        [
+          {
```
Comment (simlu) [inline test/util/sqs.spec.js]: collapse outer two

File: README.md [hunk 11]
```diff
@@ -66,6 +66,12 @@ Get the service from the underlying `aws-sdk` without initializing it. Possible
 Updates the global aws config of the underlying `aws-sdk` via `AWS.config.update`.
 In most cases this should not be necessary to use.
 
+#### sendMessageBatch(msgs: Array, queueUrl: String, options: Object = {})
+
+Splits `msgs` into groups and calls `sqs.SendMessageBatch` for every group.
```
Comment (simlu) [inline README.md]: link to aws-sdk method

File: README.md [hunk 12]
```diff
@@ -94,3 +100,23 @@ Type: `list`<br>
...
 
 Provide string list of expected AWS error codes. Promise succeeds on expected error with error code as string.
+
+### SendMessageBatch Options
+
+#### batchSize
+Type: `integer`<br>
+Default: `10`
+
+Specify the size of each batch that will be sent.
```
Comment (simlu) [inline README.md]:
```suggestion
Specify the size of each batch that will be sent. Should never exceed 10.
```

Please also place an assert that `<= 10`!

File: test/index.spec.js [hunk 13]
```diff
@@ -1,15 +1,12 @@
-const path = require('path');
 const expect = require('chai').expect;
-const nockBack = require('nock').back;
+const { desc } = require('node-tdd');
```
Comment (simlu) [inline test/index.spec.js]: please update again

**Pull request merged**